llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1430,14 +1431,30 @@ void AsmPrinter::emitFunctionBody() {
 
   // If the target wants a .size directive for the size of the function, emit
   // it.
+  const MCExpr *SizeExp = nullptr;
   if (MAI->hasDotTypeDotSizeDirective()) {
     // We can get the size as difference between the function label and the
     // temp label.
-    const MCExpr *SizeExp = MCBinaryExpr::createSub(
+    SizeExp = MCBinaryExpr::createSub(
         MCSymbolRefExpr::create(CurrentFnEnd, OutContext),
         MCSymbolRefExpr::create(CurrentFnSymForSize, OutContext), OutContext);
     OutStreamer->emitELFSize(CurrentFnSym, SizeExp);
   }
+  if (CurrentFnLocalForEH) {
+    // For CHERI exception landing pad tables we need a local alias so
+    // that we can emit non-preemptible relocations with the correct addend.
+    // See https://github.com/CTSRD-CHERI/llvm-project/issues/512.
+    // Note: We add .local so that the symbol is not deleted after assembling
+    // even though the name might start with ".L".
+    OutStreamer->emitSymbolAttribute(CurrentFnLocalForEH, MCSA_Local);
+    OutStreamer->emitSymbolAttribute(CurrentFnLocalForEH,
+                                     MCSA_ELF_TypeFunction);
```

Comment:
Needs guarding with hasDotTypeDotSizeDirective

Reply:
I looked at emitSymbolAttribute() and that function already does this check. Also it might allow emitting an appropriate directive for non-ELF too.

Reply:
Hm, guess there's redundancy then in the places I looked

Reply:
Looking more closely: WASM does the following:
`Symbol->setType(wasm::WASM_SYMBOL_TYPE_FUNCTION)`, mach-o+winCOFF simply return false and
`MCXCOFFStreamer` asserts. But I think we can probably ignore XCOFF for CHERI EH...